### PR TITLE
Integrate list and table widgets from Icinga DB Web

### PR DESCRIPTION
**Pull request by yhabteab**

### CODE REVIEW

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]
Please commit new changes separately.

**Review by sukhwinder33445 — COMMENTED** [review]
Please commit the changes from 627d7aa separately.

File: src/Common/BaseItemList.php [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+    /** @var bool Whether the list contains at least one item with an icon_image */
+    protected $hasIconImages = false;
+
+    protected $tag = 'ul';
+
+    /**
+     * Create a new item  list
+     *
+     * @param iterable<object> $data Data source of the list
```
Comment (sukhwinder33445) [inline src/Common/BaseItemList.php:36]: Same here.

File: src/Common/BaseItemTable.php [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+    /** @var iterable<object> */
+    protected $data;
+
+    protected $tag = 'ul';
+
+    /**
+     * Create a new item table
+     *
+     * @param iterable<object> $data Data source of the table
```
Comment (sukhwinder33445) [inline src/Common/BaseItemTable.php:34]: Same here.

File: src/Common/BaseItemList.php [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+    /** @var array<string, mixed> */
+    protected $baseAttributes = [
+        'class'                         => 'item-list',
+        'data-base-target'              => '_next',
+        'data-pdfexport-page-breaks-at' => '.list-item'
+    ];
+
+    /** @var iterable<object> */
+    protected $data;
```
Comment (sukhwinder33445) [inline src/Common/BaseItemList.php:26]: As described in the constructor, `$data` can also be an array.

File: src/Common/BaseItemTable.php [hunk 4]
```diff
@@ -17,20 +17,21 @@ abstract class BaseItemTable extends BaseHtmlElement
...
 
+    /** @var array<string, mixed> */
     protected $baseAttributes = [
         'class'            => 'item-table',
         'data-base-target' => '_next'
     ];
 
-    /** @var iterable */
+    /** @var iterable<object> */
     protected $data;
```
Comment (sukhwinder33445) [inline src/Common/BaseItemTable.php:27]: As described in the constructor, `$data` can also be an array.

File: src/Common/BaseTableRowItem.php [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+        $title->prepend($this->createVisual());
+
+        return $title;
+    }
+
+    /**
+     * @return ?BaseHtmlElement
+     */
+    protected function createVisual()
```
Comment (sukhwinder33445) [inline src/Common/BaseTableRowItem.php:97]: Is there a reason why you have not defined the return type here?

File: src/Common/BaseTableRowItem.php [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * Initialize the list item
+     *
+     * If you want to adjust the list item after construction, override this method.
+     */
+    protected function init(): void
+    {
+    }
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Common/BaseTableRowItem.php:115]: and here?

File: src/Widget/ShowMore.php [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    public function renderUnwrapped(): string
+    {
+        if ($this->resultSet->hasMore()) {
+            return parent::renderUnwrapped();
+        }
+
+        return '';
+    }
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Widget/ShowMore.php:55]: Return type?

File: src/Widget/Notice.php [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+     * Create a html notice
+     *
+     * @param mixed $content
+     */
+    public function __construct($content)
+    {
+        $this->content = $content;
+    }
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Widget/Notice.php:27]: Return type?

File: src/Widget/EmptyState.php [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+     * Create an empty state
+     *
+     * @param mixed $content
+     */
+    public function __construct($content)
+    {
+        $this->content = $content;
+    }
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Widget/EmptyState.php:26]: Return type?

File: src/Common/BaseStatusBar.php [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+    protected function createCount(): BaseHtmlElement
+    {
+        $total = new HtmlElement('span', Attributes::create(['class' => 'item-count']));
+
+        $this->assembleTotal($total);
+
+        return $total;
+    }
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Common/BaseStatusBar.php:40]: Return type please?

File: src/Common/BaseOrderedListItem.php [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+
+        return $this;
+    }
+
+    /**
+     * Get this element's position
+     *
+     * @return int
+     */
+    public function getOrder()
```
Comment (sukhwinder33445) [inline src/Common/BaseOrderedListItem.php:31]: Return type please?

File: src/Common/BaseOrderedItemList.php [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+use ipl\Web\Widget\EmptyState;
+
+/**
+ * @method BaseOrderedListItem getItemClass()
+ */
+abstract class BaseOrderedItemList extends BaseItemList
+{
+    protected $tag = 'ol';
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Common/BaseOrderedItemList.php:14]: Return type please?

File: src/Common/BaseListItem.php [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+    /**
+     * Initialize the list item
+     *
+     * If you want to adjust the list item after construction, override this method.
+     */
+    protected function init(): void
+    {
+    }
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Common/BaseListItem.php:154]: Return type please?

File: src/Common/BaseListItem.php [hunk 14]
```diff
@@ -0,0 +1,162 @@
...
+
+        $this->assembleTitle($title);
+
+        return $title;
+    }
+
+    /**
+     * @return ?BaseHtmlElement
+     */
+    protected function createVisual()
```
Comment (sukhwinder33445) [inline src/Common/BaseListItem.php:136]: Return type please?

File: src/Common/BaseItemTable.php [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+     * @return string
+     */
+    protected function getLayout(): string
+    {
+        return static::TABLE_LAYOUT;
+    }
+
+    abstract protected function getItemClass(): string;
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Common/BaseItemTable.php:70]: Return type please?

File: src/Common/BaseItemList.php [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * Initialize the item list
+     *
+     * If you want to adjust the item list after construction, override this method.
+     */
+    protected function init(): void
+    {
+    }
+
+    protected function assemble()
```
Comment (sukhwinder33445) [inline src/Common/BaseItemList.php:86]: Return type please?

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]
Please add the requested changes in https://github.com/Icinga/icingadb-web/pull/830.

File: src/Common/BaseOrderedListItem.php [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+        $this->order = $order;
+
+        return $this;
+    }
+
+    /**
+     * Get this element's position
+     *
+     * @return int
+     */
```
Comment (sukhwinder33445) [inline src/Common/BaseOrderedListItem.php:30]: Please add `@throws`.

File: src/Common/BaseOrderedListItem.php [hunk 18]
```diff
@@ -0,0 +1,41 @@
+<?php
+
+namespace ipl\Web\Common;
+
+use LogicException;
+
+abstract class BaseOrderedListItem extends BaseListItem
+{
+    /** @var int This element's position */
+    protected $order;
```
Comment (sukhwinder33445) [inline src/Common/BaseOrderedListItem.php:10]: Not set via constructor, so use `@var ?int`.

**Review by sukhwinder33445 — CHANGES_REQUESTED** [review]

**Review by nilmerg — DISMISSED** [review]

**Review by nilmerg — APPROVED** [review]

**Pull request merged**